lib/rust/parser/src/data/tree.rs

```diff
@@ -0,0 +1,214 @@
+//! A map-based tree for use in the macro resolver.
+
+use crate::prelude::*;
+
+
+
+// ============
+// === Tree ===
+// ============
+
+/// The type of branches in the tree.
+pub type Branches<K,V> = HashMap<K,Tree<K,V>>;
+
+/// A tree with a variable number of branches per node.
+#[derive(Derivative)]
+#[derivative(Clone)]
+#[derivative(Debug(bound="K:Debug+Eq+Hash+PartialEq, V:Debug+Eq+PartialEq"))]
+#[derivative(Default(bound="K:Eq+Hash"))]
+#[derivative(PartialEq(bound="K:Eq+Hash, V:PartialEq"))]
+#[derivative(Eq(bound="K:Eq+Hash, V:Eq"))]
+pub struct Tree<K,V> {
+    /// The value at the current node.
+    pub value : Option<V>,
+    /// The branches at the current node.
+    pub branches : Branches<K,V>
+}
+
+impl<K,V> Tree<K,V> {
+    /// Check if `self` is a leaf of the tree.
+    pub fn is_leaf(&self) -> bool {
+        self.branches.is_empty()
+    }
+
+    /// Check if `self` is a non-leaf node in the tree.
+    pub fn is_non_leaf(&self) -> bool {
+        !self.is_leaf()
+    }
+}
+
+impl<K,V> Tree<K,V> where K:Clone+Eq+Hash+PartialEq {
+    /// Create an empty tree.
+    pub fn empty() -> Self {
+        default()
+    }
+
+    /// Create a singleton tree.
+    pub fn singleton(value:V) -> Self {
+        let value    = Some(value);
+        let branches = default();
+        Self{value,branches}
+    }
+
+    /// Insert the provided `value` into the tree at the provided `path`.
+    pub fn insert(&mut self, path:&[K], value:V) {
+        if let Some((first,rest)) = path.split_first() {
+            if let Some(existing_branch) = self.branches.get_mut(first) {
+                existing_branch.insert(rest,value);
+            } else {
+                let mut new_branch = Self::empty();
+                new_branch.insert(rest,value);
+                self.branches.insert(first.clone(),new_branch);
+            }
+        } else {
+            self.value = Some(value)
+        }
+    }
+
+    /// Map the provided `f` over `self`, mutating the tree.
+    ///
+    /// This may change the value type stored in the tree.
+    ///
+    /// ## NOTE
+    /// This function is only suitable for use on trees with small depths as it is implemented in a
+    /// recursive fashion.
+    pub fn map<S,F:Copy+Fn(V) -> S>(self, f:F) -> Tree<K,S> {
+        let value         = self.value.map(f);
+        let branches_iter = self.branches.into_iter().map(|(k,v)| (k,v.map(f)));
+        let branches      = branches_iter.collect::<Branches<K,S>>();
+        Tree{value,branches}
```

Comment:
USe Self! Less refactoring in the future :)

Reply:
You can't use self as this is a type changing operation.